src/java.base/share/classes/java/io/ObjectOutputStream.java

```diff
@@ -2026,8 +2027,46 @@ public void writeChars(String s) throws IOException {
             }
         }
 
-        public void writeUTF(String s) throws IOException {
-            writeUTF(s, getUTFLength(s));
+        public void writeUTF(String str) throws IOException {
+            writeUTFInternal(str, false);
+        }
+
+        private void writeUTFInternal(String str, boolean writeHeader) throws IOException {
+            int strlen = str.length();
+            int countNonZeroAscii = JLA.countNonZeroAscii(str);
+            int utflen = utfLen(str, countNonZeroAscii);
+            if (utflen <= 0xFFFF) {
+                if(writeHeader) {
+                    writeByte(TC_STRING);
+                }
+                writeShort(utflen);
+            } else {
+                if(writeHeader) {
+                    writeByte(TC_LONGSTRING);
+                }
+                writeLong(utflen);
```

Comment:
The old plain `writeUTF` does not write a long utf if the length is too long, but rather throws an `UTFDataFormatException`. There should be code paths that no longer throw this exception, because a few tests, including:
- `java/io/Serializable/longString/LongString.java`
- `java/io/Serializable/sanityCheck/SanityCheck.java`
- `java/text/Format/DateFormat/DateFormatRegression.java`
are failing. I think this might be the root cause.

Reply:
The reason for the error is that when the `drain` method is called, the local variable pos is not updated to the field `pos`